url64d634b8f0_agdturner_generic_tgz-pJ8-uk_ac_leeds_ccg_andyt_generic_visualisation_charts_Generic_BarChartJ8/src/uk/ac/leeds/ccg/andyt/generic/utilities/Generic_Execution.java

```diff
@@ -0,0 +1,165 @@
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import uk.ac.leeds.ccg.andyt.generic.visualisation.Generic_Visualisation;
+
+/**
+ * Generic utility class for execution handling.
+ */
+public class Generic_Execution {
+
+    /**
+     * Uses default delays 
+     * @param executorService
+     * @param futures
+     * @param obj 
+     */
+    public static void shutdownExecutorService(
+            ExecutorService executorService,
+            HashSet<Future> futures,
+            Object obj) {
+        long delay = 1000;
+        long maxWait = 10;
+        shutdownExecutorService(
+                executorService,
+                futures,
+                obj,
+                delay,
+                maxWait);
+    }
+
+    /**
+     * @param executorService
+     * @param futures
+     * @param obj
+     * @param delay Number of milliseconds for loop to wait before testing if a
+     * Future is returned.
+     * @param maxWait Number of minutes to block shutdown after all futures 
+     * returned 
+     */
+    public static void shutdownExecutorService(
+            ExecutorService executorService,
+            HashSet<Future> futures,
+            Object obj,
+            long delay,
+            long maxWait) {
+        //What is still left to do from futures?
+        Iterator<Future> ite = futures.iterator();
+        System.out.println("There are " + futures.size() + " jobs to check.");
+        int doneJobsCounter = 0;
+        int notDoneJobsCounter = 0;
+        while (ite.hasNext()) {
+            Future f = ite.next();
+            if (f.isDone()) {
+                doneJobsCounter++;
+            } else {
+                notDoneJobsCounter++;
+            }
+        }
+        System.out.println("There are " + doneJobsCounter + " jobs done.");
+        System.out.println("There are " + notDoneJobsCounter + " jobs not done.");
+        ite = futures.iterator();
+        while (ite.hasNext()) {
+            Future future = ite.next();
+            long counter = 0;
+            while (!future.isDone()) {
+                System.out.println("Job not done waiting " + delay + " milliseconds having already waited " + counter + " milliseconds");
+                counter += delay;
+                //try {
+                Generic_Execution.waitSychronized(obj, delay);
+                //} catch (InterruptedException e) {
+                //    e.printStackTrace();
+                //}
+            }
+        }
+        System.out.println("Jobs done.");
+        executorService.shutdown();
+
+        try {
+            boolean allterminated = executorService.awaitTermination(maxWait, TimeUnit.MINUTES);
+            System.out.println("All output terminated " + allterminated);
+        } catch (InterruptedException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)